CMSgov/macpro-quickstart-serverless
# Use IAM roles per function

## Purpose

One of the ways to mitigate [security risks for serverless applications](https://github.com/puresec/sas-top-10) is to use the principle of least privilege for function permissions. When a default IAM role is defined for a serverless service, each function receives all the permissions defined for the default role.  This can lead to overly broad function permissions, which violates the principle.  Each function should be granted the minimum set of permissions required to complete the job of the function.

This change removes all default lambda execution roles and adds per-function roles with the minimum permissions to execute the function handlers, and adds a section to the `services` README that explains this as a best practice for serverless security. 

#### Linked Issues to Close

N/A

## Approach

It's possible to define [custom IAM roles](https://www.serverless.com/framework/docs/providers/aws/guide/iam#custom-iam-roles), but if you do so you need to manually handle the details of defining IAM permissions for logs and stream events that are usually abstracted away by the Serverless framework.

This change uses the [serverless-iam-roles-per-function](https://www.serverless.com/plugins/serverless-iam-roles-per-function) plugin, which allows defining custom roles for each function while still abstracting away the permissions that are required for all serverless functions.  This plugin is recommended in the [Serverless documentation for IAM permissions](https://www.serverless.com/framework/docs/providers/aws/guide/iam#custom-iam-roles-for-each-function).

## Learning
- https://www.serverless.com/plugins/serverless-iam-roles-per-function
- https://www.serverless.com/framework/docs/providers/aws/guide/iam
- https://github.com/puresec/sas-top-10


## Assorted Notes/Considerations

Manual testing: 

I deployed each service after adding per-function IAM roles, and tested each function to confirm that it still had the required IAM permissions to execute without error. 

#### Pull Request Creator Checklist


#### Pull Request Reviewer/Assignee Checklist
